lib/usd/translators/meshWriter_BlendShapes.cpp

```diff
@@ -0,0 +1,977 @@
+        const size_t        numOrigOffsets = origOffsetsArray.size();
+        TF_VERIFY(numOrigOffsets != 0);
+        VtVec3fArray& newOffsetsArray = unionOffsetsArrays[i];
+        newOffsetsArray.assign(numUnionIndices, GfVec3f(0.0f));
+
+        const VtVec3fArray& origNormalsArray = normalsArrays[i];
+#ifdef _DEBUG
+        const size_t numOrigNormals = origNormalsArray.size();
+        TF_VERIFY(numOrigOffsets == numOrigNormals);
+#endif
+        VtVec3fArray& newNormalsArray = unionNormalsArrays[i];
+        newNormalsArray.assign(numUnionIndices, GfVec3f(0.0f));
+
+        const VtIntArray& origIndicesArray = indicesArrays[i];
+        for (size_t j = 0, k = 0; j < numUnionIndices; ++j) {
+            int index = unionIndices[j];
+            int origIndex = origIndicesArray[k];
+            if (index != origIndex || k > numOrigOffsets - 1) {
+                GfVec3f sentinel(0.0f);
+                newOffsetsArray[j] = sentinel;
+                newNormalsArray[j] = sentinel;
+            } else {
+                newOffsetsArray[j] = origOffsetsArray[k];
+                newNormalsArray[j] = origNormalsArray[k];
+                ++k;
+            }
+        }
+    }
+
+    return;
+}
+
+// NOTE: (yliangsiew) This gets called once for each shape being exported
+// under a single transform.
+MObject PxrUsdTranslators_MeshWriter::writeBlendShapeData(UsdGeomMesh& primSchema)
+{
+    MStatus                    stat;
+    const UsdMayaJobExportArgs exportArgs = this->_GetExportArgs();
+
+    MDagPath deformedMeshDagPath = this->GetDagPath();
+    MObject  deformedMesh = deformedMeshDagPath.node();
+    TF_VERIFY(deformedMesh.hasFn(MFn::kMesh) == true);
+
+    MObjectArray blendShapeDeformers;
+    MIntArray    blendShapeCxnIndices;
+
+    // TODO: (yliangsiew) Figure out if this can be isolated. It's kind of hard
+    // because we want to avoid repeated walks through the DG.
+    std::vector<MayaBlendShapeDatum> blendShapeDeformerInfos;
+    stat = mayaGetBlendShapeInfosForMesh(deformedMesh, blendShapeDeformerInfos);
+    if (stat != MStatus::kSuccess) {
+        TF_WARN(
+            "Could not read blendshape information for the mesh: %s.",
+            deformedMeshDagPath.fullPathName().asChar());
+        return MObject::kNullObj;
+    }
+
+    size_t numOfBlendShapeDeformers = blendShapeDeformerInfos.size();
+    switch (numOfBlendShapeDeformers) {
+    case 0:
+        TF_WARN(
+            "Cannot find any blendshape deformers for the mesh: %s",
+            deformedMeshDagPath.fullPathName().asChar());
+        return MObject::kNullObj;
+    case 1: break;
+    default:
+        // TODO: (yliangsiew) For multiple blend shape deformers, what do we do?
+        // Do we collapse the shapes from multiple blend targets together, or just
+        // write out only the "closest" blendshape deformer's targets? Or just write all
+        // of them and print this warning to end-users?
```

Comment:
I think eventually we'll want a 1 to 1 mapping between deformer nodes in Maya and UsdSkel compute items written to USD. Doing a special case to merge two blend shapes in a row won't help when there's a wrap deformer between them. We'll eventually have to support writing a graph of UsdSkel compute items to the USD file.

I think this change, even though it has a bunch of limitations, is a good step towards that final goal.